### PR TITLE
Lite wheel optimization

### LINKED ISSUE #7853: Optimize the wheel file contents for Lite
The following files included in the wheel are not needed for Lite.
* `gradio/_frontend_code`
* `gradio/templates`
* `gradio/test_data`
* `gradio/node`
* `gradio/*.pyi`

### PR DESCRIPTION
**Pull request by whitphx**
## Description

* Fix the Lite wheel build to be lighter for Lite (thanks to #7858 by @freddyaboulton )
* Fix the CI workflow so we can check the built files before publishing.

Resolves #7853

### CODE REVIEW

**Comment by whitphx** [comment]
Found that this step triggers building the package, including copying the `_frontend_code`
https://github.com/gradio-app/gradio/blob/72661e3391a432ccd77fb3100e69fbdf95093931/.github/actions/install-all-deps/action.yml#L77-L83

**Comment by whitphx** [comment]
The functional test failed because this `hatch` started to be needed.
I confirmed it would be working after merged into main. Ref: 571823b, https://github.com/gradio-app/gradio/actions/runs/8477748173/job/23229061320?pr=7855

File: .github/workflows/publish.yml
Comment (whitphx) [inline .github/workflows/publish.yml:35]: It might be better if the NPM package created in this step will be published in the following step as it's guaranteed that the NPM-published file is exactly same as what we can check from the uploaded artifacts?

File: .github/workflows/publish.yml
Comment (whitphx) [inline .github/workflows/publish.yml:6]: @pngwn Do you think it's ok to run this workflow on each PR as we discussed?

**Comment by abidlabs** [comment]
@whitphx is this ready to be reviewed or still a draft?

**Comment by whitphx** [comment]
@abidlabs Oh yes it's to be reviewed thanks, but not ready to merge: I want opinions about the comments above.

File: .github/workflows/publish.yml
Comment (pngwn) [reply .github/workflows/publish.yml:6]: No, we can only run the publish when we want to trigger the publish workflow, this will also fail for third part contributions as they won't have access to secrets. We need to expose the assets in one of the other workflows, the deploy space workflow is probably the most appropriate.

File: .github/workflows/publish.yml
Comment (pngwn) [reply .github/workflows/publish.yml:35]: Aren't we already building part of this in the `install-deps` workflow above? 

We should keep the builds together where possible, I think moving this tot he install deps action would make the most sense. Then we can have access to these assets in a variety of workflows if needed.

**Review by pngwn — CHANGES_REQUESTED** [review]
Left a few comments that we need to address before we can merge this.

File: .github/workflows/publish.yml
Comment (whitphx) [reply .github/workflows/publish.yml:35]: Oh the `install-frontend-deps` action has this part building Lite 👀  
You mean we should move `pnpm pack` to this?
https://github.com/gradio-app/gradio/blob/024b44c553e73f1d6c00d4c8cf12ae38181a454d/.github/actions/install-frontend-deps/action.yml#L54-L60

(Is it ok that the "install-*-deps" action is also responsible for building things which sounds not intuitive?)

File: .github/workflows/publish.yml
Comment (whitphx) [reply .github/workflows/publish.yml:6]: > We need to expose the assets in one of the other workflows, the deploy space workflow is probably the most appropriate.

OK, so you mean
1. to move `pnpm pack` to the install-deps action
2. and to use it from deploy-spaces workflow then expose the artifacts
?
(Same question: is it ok to add such steps to the deploy-space workflow without renaming it?)

Comment (pngwn) [reply .github/workflows/publish.yml]: Yep. And yeah it's fine to add it to the deploy space workflow without renaming.

**Comment by whitphx** [comment]
@pngwn Like this?

File: .github/workflows/publish.yml
Comment (whitphx) [reply .github/workflows/publish.yml]: Thanks, so I applied the changes but `pnpm pack` is not added to the install-deps action as it's now only needed in one place.

**Comment by whitphx** [comment]
```
InvalidRequirement: Invalid URL: {root:uri}/../.config/lite-builder
 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed.
ERROR: "pybuild:gradio" exited with 1.
 ELIFECYCLE  Command failed with exit code 1.
/home/runner/work/gradio/gradio/js/app:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @gradio/app@1.30.1 build:lite: `pnpm pybuild && pnpm cssbuild && pnpm --filter @gradio/client build && pnpm --filter @gradio/wasm build && vite build --mode production:lite`
Exit status 1
Error: Process completed with exit code 1.
```

This error occurs both on local and CI.

**Comment by whitphx** [comment]
Specifying `packaging>=23.2` doesn't work because `23.2` will be automatically installed by Pyodide as [it ships the version](https://pyodide.org/en/stable/usage/packages-in-pyodide.html) so it causes version conflicts.

**Comment by whitphx** [comment]
Confirmed the built wheel was uploaded and it worked in local.

**Comment by whitphx** [comment]
The functional test failed because this hatch started to be needed.
I confirmed it would be working after merged into main. Ref: https://github.com/gradio-app/gradio/commit/571823b4a05f7e41e0b3731d51c5bd86b2e17ddc, https://github.com/gradio-app/gradio/actions/runs/8477748173/job/23229061320?pr=7855

**Comment by freddyaboulton** [comment]
I think we need to modify the install action to install hatch for the lite build?

**Comment by whitphx** [comment]
@freddyaboulton Thanks,
is it this?
https://github.com/gradio-app/gradio/pull/7855/files#diff-704d0432232d65a5a81503f0e97f31053a3826e4e291fd3920eb24d1bb5d4e3aR50

**Review by freddyaboulton — APPROVED** [review]
Lovely PR @whitphx ! Excited to ship a smaller wheel. Left a small comment about the parameters you removed from `install-frontend-deps`. Otherwise this is good to merge.

File: .github/actions/install-frontend-deps/action.yml
Comment (whitphx) [reply .github/actions/install-frontend-deps/action.yml:8]: Oops missed them, thanks!

File: .github/workflows/publish.yml
Comment (whitphx) [inline .github/workflows/publish.yml:28]: This hatch's issue was solved and the fix has already been released: https://github.com/pypa/hatch/releases/tag/hatchling-v1.23.0
So this `packaging` spec can be removed.

Comment (whitphx) [reply .github/workflows/publish.yml]: hmm, it didn't work well.
https://github.com/gradio-app/gradio/actions/runs/8828922391/job/24238872108

**Review by freddyaboulton — COMMENTED** [review]
Looks good to merge @whitphx !

**Comment by whitphx** [comment]
The functional test fails because the `pnpm pybuild` command used in this PR is from this branch and uses hatch, but the CI file is from main and doesn't install hatch.
I will add `pip install hatch` to the `pybuild` command directly to make it pass temporarily in this PR.

**Comment by whitphx** [comment]
@pngwn Can you approve this PR if you are ok because the merge is blocked?

File: .github/workflows/publish.yml
Comment (whitphx) [reply .github/workflows/publish.yml:28]: Keep installing `packaging` here then.

**Comment by whitphx** [comment]
Removed some steps from the `deploy-spaces` workflow as the same steps are done in the `install-all-deps` action, as @pngwn suggested.
Looks like the workflow ran correctly: https://github.com/gradio-app/gradio/actions/runs/8887216886/job/24402132060

**Pull request merged**
